onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -66,6 +66,82 @@ TensorProto ToTensor<onnxruntime::BFloat16>(const std::vector<onnxruntime::BFloa
   return t;
 }
 
+template <>
+TensorProto ToTensor<onnxruntime::Float8E4M3FN>(const onnxruntime::Float8E4M3FN& value) {
+  TensorProto t;
+  t.set_data_type(TensorProto_DataType_FLOAT8E4M3FN);
+  t.add_int32_data(value.val);
+  return t;
+}
+
+template <>
+TensorProto ToTensor<onnxruntime::Float8E4M3FNUZ>(const onnxruntime::Float8E4M3FNUZ& value) {
+  TensorProto t;
+  t.set_data_type(TensorProto_DataType_FLOAT8E4M3FNUZ);
+  t.add_int32_data(value.val);
+  return t;
+}
+
+template <>
+TensorProto ToTensor<onnxruntime::Float8E5M2>(const onnxruntime::Float8E5M2& value) {
+  TensorProto t;
+  t.set_data_type(TensorProto_DataType_FLOAT8E5M2);
+  t.add_int32_data(value.val);
+  return t;
+}
+
+template <>
+TensorProto ToTensor<onnxruntime::Float8E5M2FNUZ>(const onnxruntime::Float8E5M2FNUZ& value) {
+  TensorProto t;
+  t.set_data_type(TensorProto_DataType_FLOAT8E5M2FNUZ);
+  t.add_int32_data(value.val);
+  return t;
+}
+
+template <>
+TensorProto ToTensor<onnxruntime::Float8E4M3FN>(const std::vector<onnxruntime::Float8E4M3FN>& values) {
+  TensorProto t;
+  t.clear_int32_data();
+  t.set_data_type(TensorProto_DataType_FLOAT8E4M3FN);
+  for (const onnxruntime::Float8E4M3FN& val : values) {
+    t.add_int32_data(val.val);
+  }
```

Comment:
Apart from the enum, the code is repeated. Would be possible to generalize and re-use it?

Reply:
The above you can generalize to a function that takes data type and uint16_t.
The one that takes a vector, you can generalize to take the datatype and a span of uint16_t. All of them can be cast to const uint16_t*.

Reply:
The template ToTensor<> seems to be declared in onnx and overloaded in onnxruntime. I tried to factorize as much as I could but I cannot get rid of the template.